pkg/controllers/machinesync/machine_sync_controller_test.go

```diff
@@ -738,6 +765,258 @@ var _ = Describe("With a running MachineSync Reconciler", func() {
+				"machine.openshift.io/cluster-api-cluster":      "ci-op-gs2k97d6-c9e33-2smph",
+				"machine.openshift.io/cluster-api-machine-role": "worker",
+				"machine.openshift.io/cluster-api-machine-type": "worker",
+				"machine.openshift.io/cluster-api-machineset":   "ci-op-gs2k97d6-c9e33-2smph-worker-us-west-2b",
+				"machine.openshift.io/instance-type":            "m6a.xlarge",
+				"cluster.x-k8s.io/cluster-name":                 "ci-op-gs2k97d6-c9e33-2smph",
+				"cluster.x-k8s.io/set-name":                     "ci-op-gs2k97d6-c9e33-2smph-worker-us-west-2b",
+				"node-role.kubernetes.io/worker":                "",
+			}).WithAnnotations(map[string]string{
+				"machine.openshift.io/instance-state": "running",
+			}).Build()
+			Expect(k8sClient.Create(ctx, capiMachine)).Should(Succeed())
+
+			// Status controller doesn't run in envtest, its part of KCM - we've got to sleep?
+
+			// 'On my machine' 1s seems to be about right for the time required
+			// for the VAP to be loaded into the API Server
+			time.Sleep(1 * time.Second)
+
+		})
+
+		AfterEach(func() {
+			By("Removing the VAP and its VAP binding")
+			Expect(k8sClient.Delete(ctx, policyBinding)).Should(Succeed())
+			Expect(k8sClient.Delete(ctx, policy)).Should(Succeed())
+		})
+
+		Context("with status.AuthoritativeAPI: Machine API", func() {
+			BeforeEach(func() {
+				By("Setting the MAPI machine AuthoritativeAPI to Machine API")
+				Eventually(k.UpdateStatus(mapiMachine, func() {
+					mapiMachine.Status.AuthoritativeAPI = machinev1beta1.MachineAuthorityMachineAPI
+				})).Should(Succeed())
+			})
+
+			It("updating the spec should be allowed", func() {
+				Eventually(k.Update(mapiMachine, func() {
+					mapiMachine.Spec.ObjectMeta.Labels = map[string]string{"foo": "bar"}
+				}), timeout).Should(Succeed(), "expected success when updating the spec")
+			})
+
+		})
+
+		Context("with status.AuthoritativeAPI: Cluster API", func() {
+			BeforeEach(func() {
+				By("Setting the MAPI machine AuthoritativeAPI to Cluster API")
+				Eventually(k.UpdateStatus(mapiMachine, func() {
+					mapiMachine.Status.AuthoritativeAPI = machinev1beta1.MachineAuthorityClusterAPI
+				})).Should(Succeed())
+			})
+
+			It("updating the spec (outside of authoritative api) should be prevented", func() {
+				var err error
+				Eventually(func() error {
+					err = k.Update(mapiMachine, func() {
+						mapiMachine.Spec.ObjectMeta.Labels = map[string]string{"foo": "bar"}
+					})()
+
+					return err
+				}, timeout).ShouldNot(Succeed())
+				expectVAPError(err, "You may only modify spec.authoritativeAPI")
+			})
+
+			It("updating the spec.authoritativeAPI should be allowed", func() {
+				Eventually(k.Update(mapiMachine, func() {
+					mapiMachine.Spec.AuthoritativeAPI = machinev1beta1.MachineAuthorityMachineAPI
+				}), timeout).Should(Succeed(), "expected success when updating spec.authoritativeAPI")
+			})
+
+			Context("when trying to update metadata.labels", func() {
```

Comment:
Test a label where the original value is `""` and one where you take a non empty value and try to change it to `""`

Reply:
Using the new `variables.oldLabels[?k].orValue(null)` style (with null) I think fixes all of these issues. 

Added tests anyway.